src/System.Private.Uri/src/System/IPv6AddressHelper.cs

```diff
@@ -61,20 +63,20 @@ internal static unsafe string ParseCanonicalName(string str, int start, ref bool
                 }
 
                 // Compression; 1::1, ::1, 1::
-                if (range.Key == i)
+                if (rangeStart == i)
                 {
                     // Start compression, add :
                     stackSpace[pos++] = ':';
                 }
 
-                if (range.Key <= i && range.Value == (NumberOfLabels - 1))
+                if (rangeStart <= i && rangeEnd == NumberOfLabels)
                 {
                     // Remainder compressed; 1::
                     stackSpace[pos++] = ':';
                     break;
                 }
 
-                if (range.Key <= i && i <= range.Value)
+                if (rangeStart <= i && i < rangeEnd)
```

Comment:
Why does the range end change by 1?

Reply:
It came from two semantically incompatible versions of `FindCompressionRange` - I'll clean that up a bit and provide some more comments.